File: plugins/app/qwenpaw-creator/ui/src/contracts/creator/commands.ts
import type { ExpectedTargetVersion } from './common';

export type CreatorCommandType =
  | 'GENERATE_SCRIPT'
  | 'IMPORT_SCRIPT'
  | 'SET_STRATEGY_TEXT'
  | 'SET_SECTION_TEXT'
  | 'SET_UNIT_TEXT'
  | 'PLAN_UNITS'
  | 'CREATE_SECTION'
  | 'DELETE_SECTION'
  | 'MOVE_SECTION'
  | 'CREATE_UNIT'
  | 'DELETE_UNIT'
  | 'MOVE_UNIT'
  | 'CHANGE_UNIT_ROUTE'
  | 'UPSERT_SHOT'
  | 'DELETE_SHOT'
  | 'MOVE_SHOT'
  | 'BIND_REFERENCE'
  | 'UNBIND_REFERENCE'
  | 'GENERATE_STORYBOARD_PROMPT'
  | 'GENERATE_STORYBOARD_IMAGE'
  | 'GENERATE_VIDEO_PROMPT'
  | 'GENERATE_R2V_VIDEO'
  | 'BUILD_EDIT_PLAN'
  | 'EXECUTE_EDIT'
  | 'SET_EDIT_CLIP_RANGE'
  | 'SET_EDIT_CLIP_OS'
  | 'SET_EDIT_CLIP_TRANSITION'
  | 'MOVE_EDIT_CLIP'
  | 'DELETE_EDIT_CLIP'
  | 'SET_EDIT_AUDIO_PLAN'
  | 'ATTACH_SOURCE_ASSETS'
  | 'DETACH_SOURCE_ASSETS'
  | 'SUPPLEMENT_ASSET'
  | 'GENERATE_ASSET'
  | 'SELECT_ARTIFACT_VERSION'
  | 'SET_SECTION_COMPOSE_SELECTION'
  | 'SET_SECTION_COMPOSE_TRANSITION'
  | 'STITCH_SECTION'
  | 'SET_FINAL_COMPOSE_SELECTION'
  | 'SET_FINAL_COMPOSE_TRANSITION'
  | 'COMPOSE_FINAL_VIDEO'
  | 'ANALYZE_SOURCE_MEDIA';

export interface CreatorCommand {
  clientCommandId: string;
  editSessionId?: string;
  type: CreatorCommandType;
  targetRef: string;
  arguments: Record<string, unknown>;
  context?: Record<string, unknown>;
  expectedTargetVersions: ExpectedTargetVersion[];
  expectedPresentationVersion?: string;
  expectedDecisionToken?: string;
  expectedOverlayHead?: string;
}

export interface CreatorCommandAccepted {
  commandId: string;
  status:
    | 'RECEIVED'
    | 'APPLIED'
    | 'QUEUED'
    | 'DEFERRED_UNTIL_REVIEW_RESOLVED'
    | 'CONSUMED'
    | 'REJECTED';
  eventSeq: number;
  transactionId?: string;
  messageSeq?: number;
  workingHead?: string;
}

File: plugins/app/qwenpaw-creator/backend/schemas/commands.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from domain.enums import CreatorCommandType

from .common import ExpectedTargetVersion, StrictModel


class CreatorCommandRequest(StrictModel):
    client_command_id: str = Field(alias="clientCommandId")
    edit_session_id: str | None = Field(None, alias="editSessionId")
    type: CreatorCommandType
    target_ref: str = Field(alias="targetRef")
    arguments: dict[str, Any] = Field(default_factory=dict)
    context: dict[str, Any] = Field(default_factory=dict)
    expected_target_versions: list[ExpectedTargetVersion] = Field(default_factory=list, alias="expectedTargetVersions")
    expected_presentation_version: str | None = Field(None, alias="expectedPresentationVersion")
    expected_decision_token: str | None = Field(None, alias="expectedDecisionToken")
    expected_overlay_head: str | None = Field(None, alias="expectedOverlayHead")


class CreatorCommandAccepted(StrictModel):
    command_id: str = Field(alias="commandId")
    status: Literal[
        "RECEIVED",
        "APPLIED",
        "QUEUED",
        "DEFERRED_UNTIL_REVIEW_RESOLVED",
        "CONSUMED",
        "REJECTED",
    ]
    event_seq: int = Field(alias="eventSeq", ge=0)
    transaction_id: str | None = Field(None, alias="transactionId")
    message_seq: int | None = Field(None, alias="messageSeq", ge=1)
    working_head: str | None = Field(None, alias="workingHead")
